Lib/asyncio/locks.py

```diff
@@ -412,3 +412,212 @@ def release(self):
         if self._value >= self._bound_value:
             raise ValueError('BoundedSemaphore released too many times')
         super().release()
+
+
+class Barrier(mixins._LoopBoundMixin):
+    """Asyncio equivalent to threading.Barrier
+
+    Implements a Barrier.
+    Useful for synchronizing a fixed number of tasks at known synchronization
+    points.  Tasks block on 'wait()' and are simultaneously awoken once they
+    have all made their call.
+    """
+
+    def __init__(self, parties, action=None):
+        """Create a barrier, initialised to 'parties' tasks.
+        'action' is a coroutine which, when supplied, will be called by
```

Comment:
```suggestion
        """Create a barrier, initialized to 'parties' tasks.
        
        'action' is an awaitable which, when supplied, will be called by
```

Reply:
Actually, this is an interesting question -- a barrier can be reused multiple times, right?

If so, the `action` coroutine could be awaited multiple times too, right? If so, it should be `action_factory` and be a callable that returns an awaitable.

Reply:
I'm pretty sure that async code should use an async callback.
The `action` argument removal is an option. It could be added later on-demand, the addition is backward compatible.

Reply:
> I'm pretty sure that async code should use an async callback

Ok

I think there is a solution: `action` must be a `coroutinefunction`, even if it is included in a `functools.partial` call.  The type of `action` will be checked in the `__init__` method.
Now the `action` parameter is a callable, returning an awaitable. The call to `action` becomes `await self._action()`

What do you think of this ?

Reply:
+1 for dropping.

Moreover, since the waiting order is deterministic in `asyncio.Condition`, the first awaited task wakes up first before all others.

Reply:
+1 for dropping.
I remove `action` parameter

Reply:
Should we move `BrokenBarrierError` exception from 'asyncio.locks.py'  to 'asyncio.exceptions.py' file ?

Reply:
> Should we move BrokenBarrierError exception...

I guess it would be nice but not strictly necessary.
`exceptions.py` reduces cross-imports. I expect that `BrokenBarrierError` is never used outside of the `Barrier` class itself.
On the other hand, putting all classes together could be good for source side readers.

Reply:
I 'd just seen you made modifications in python files. 
Could you give me time please to read what tou did ? Thank

Reply:
Yes, sure. 
No rush, please.